nowcasting_dataloader/data_sources/opticalflow/opticalflow_model.py

```diff
@@ -0,0 +1,106 @@
+""" Model for output of satellite data """
+from __future__ import annotations
+
+import logging
+from typing import Optional
+
+import numpy as np
+import xarray as xr
+from pydantic import Field
+
+from nowcasting_dataloader.data_sources.datasource_output import Array, DataSourceOutputML
+from nowcasting_dataloader.xr_utils import re_order_dims
+
+logger = logging.getLogger(__name__)
+
+# TODO update
+OPTICALFLOW_MEAN = {
+    # "HRV": 236.13257536395903,
+    "IR_016": 291.61620182554185,
+    "IR_039": 858.8040610176552,
+    "IR_087": 738.3103442750336,
+    "IR_097": 773.0910794778366,
+    "IR_108": 607.5318145165666,
+    "IR_120": 860.6716261423857,
+    "IR_134": 925.0477987594331,
+    "VIS006": 228.02134593063957,
+    "VIS008": 257.56333202381205,
+    "WV_062": 633.5975770915588,
+    "WV_073": 543.4963868823854,
+}
+
+OPTICALFLOW_STD = {
+    # "HRV": 935.9717382401759,
+    "IR_016": 172.01044433112992,
+    "IR_039": 96.53756504807913,
+    "IR_087": 96.21369354283686,
+    "IR_097": 86.72892737648276,
+    "IR_108": 156.20651744208888,
+    "IR_120": 104.35287930753246,
+    "IR_134": 104.36462050405994,
+    "VIS006": 150.2399269307514,
+    "VIS008": 152.16086321818398,
+    "WV_062": 111.8514878214775,
+    "WV_073": 106.8855172848904,
+}
+
+
+class OpticalFlowML(DataSourceOutputML):
+    """Model for output of satellite data"""
+
+    # Shape: [batch_size,] seq_length, width, height, channel
+    data: Array = Field(
+        ...,
+        description="Optical flow images. Shape: [batch_size,] seq_length, width, height, channel",
+    )
+    x: Array = Field(
+        ...,
+        description="The x (OSGB geo-spatial) coordinates of the optical flow images. "
+        "Shape: [batch_size,] width",
+    )
+    y: Array = Field(
+        ...,
+        description="The y (OSGB geo-spatial) coordinates of the optical flow images. "
+        "Shape: [batch_size,] height",
+    )
+
+    time: Array = Field(
+        ...,
+        description="Time index of optical flow data at 5 minutes past the hour {0, 5, ..., 55}. "
+        "*not* the {4, 9, ..., 59} timings of the optical flow imagery. "
+        "Datetimes become Unix epochs (UTC) represented as int64 just before being"
+        "passed into the ML model.",
+    )
+
+    channels: Optional[Array] = Field(
+        list(OPTICALFLOW_MEAN.keys()), description="List of the optical flow channels"
+    )
+
+    def get_datetime_index(self) -> Array:
+        """Get the datetime index of this data"""
+        return self.time
+
+    @staticmethod
+    def from_xr_dataset(xr_dataset: xr.Dataset):
+        """Change xr dataset to model."""
+
+        # make sure the dims are in the correct order
+        xr_dataset = re_order_dims(xr_dataset)
+
+        # convert to torch dictionary
+        opticalflow_batch_ml = xr_dataset.torch.to_tensor(["data", "time", "x", "y"])
+
```

Comment:
need to load channels to, and then TODO can be removed

Reply:
Actually im not sure. I found a problem when normalizing if the data didnt have all the channels. Ill add a test

Reply:
Yeah, you have to set the channels directly for it to work without all the satellite channels. I did it for the satellite data source , in loading the batch I think, just need to find it

Reply:
I just added it, basically when loading from the Batch to BatchML, you set the channels to the available channels from the xarray dataset